tensorflow/contrib/opt/python/training/model_average_optimizer.py

```diff
@@ -0,0 +1,696 @@
+# 2017 Contrib.
+# ==============================================================================
+
+"""Synchronize replicas for model average training."""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import time
+import six
+
+from tensorflow.core.framework import node_def_pb2
+from tensorflow.python.framework import device as pydev
+from tensorflow.python.training import server_lib
+from tensorflow.python.training import device_setter
+from tensorflow.core.framework import types_pb2
+from tensorflow.python.framework import ops
+from tensorflow.python.ops import array_ops
+from tensorflow.python.ops import control_flow_ops
+from tensorflow.python.ops import data_flow_ops
+from tensorflow.python.ops import init_ops
+from tensorflow.python.ops import math_ops
+from tensorflow.python.ops import state_ops
+from tensorflow.python.ops import variables
+from tensorflow.python.ops import variable_scope
+from tensorflow.python.platform import tf_logging as logging
+from tensorflow.python.training import queue_runner
+from tensorflow.python.training import session_manager
+from tensorflow.python.training import session_run_hook
+
+class _ModelAverageDeviceChooser(object):
+  """Class to choose devices for Ops in a model average training setup.
+
+  This class is not to be used directly by users.  See instead
+  `model_average_device_setter()` below.
+  """
+
+  def __init__(self, ps_tasks, ps_device, worker_device, ps_ops,
+               ps_strategy):
+    """Create a new `_ReplicaDeviceChooser`.
+
+    Args:
+      ps_tasks: Number of tasks in the `ps` job.
+      ps_device: String.  Name of the `ps` job.
+      worker_device: String.  Name of the `worker` job.
+      ps_ops: List of strings representing `Operation` types that need to be
+        placed on `ps` devices.
+      ps_strategy: A callable invoked for every ps `Operation` (i.e. matched by
+        `ps_ops`), that takes the `Operation` and returns the ps task index to
+        use.
+    """
+    self._ps_tasks = ps_tasks
+    self._ps_device = ps_device
+    self._worker_device = worker_device
+    self._ps_ops = ps_ops
+    self._ps_strategy = ps_strategy
+
+  def device_function(self, op):
+    """Choose a device for `op`.
+
+    Args:
+      op: an `Operation`.
+
+    Returns:
+      The device to use for the `Operation`.
+    """
+
+    current_device = pydev.DeviceSpec.from_string(op.device or "")
+
+    node_def = op if isinstance(op, node_def_pb2.NodeDef) else op.node_def
+    # TODO(chen meng): only global variables should be placed on ps. Now we just
+    # do this according to substring in var name.
+    if self._ps_tasks and self._ps_device and node_def.op in self._ps_ops \
+       and node_def.name.find('modelAverage') != -1:
```

Comment:
I have tried it, but it can't work because this is in device_setter() function and device_setter() is called before add_to_collection() when construct a variables. If any other idea?